src/qibo/backends/numpy.py

```diff
@@ -398,7 +398,8 @@ def execute_circuit(self, circuit, initial_state=None, nshots=1000):
             else:
                 return self.execute_circuit(initial_state + circuit, None, nshots)
         elif initial_state is not None:
-            initial_state = self.cast(initial_state)
+            if initial_state.dtype != self.dtype:
+                initial_state = self.cast(initial_state)
```

Comment:
Is this needed in the `numpy` backend?

Reply:
If you use a float/int for the initial state it will be casted to complex which is the correct dtype for the computation. Moreover this function is also used by the pytorch backend, so this will make the circuit run even if you are passing a numpy array as initial state when using the pytorch backend.